cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/co/cask/cdap/etl/batch/spark/ETLSpark.java

```diff
@@ -0,0 +1,179 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *
+ */
+public class ETLSpark extends AbstractSpark {
+
+  private static final Gson GSON = new Gson();
+
+  private final ETLBatchConfig config;
+
+  private List<Finisher> finishers;
+//  private BatchConfigurable<BatchSourceContext> batchSource;
+//  private List<BatchConfigurable<BatchSinkContext>> batchSinks;
+  private List<File> cleanupFiles;
+
+  public ETLSpark(ETLBatchConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  protected void configure() {
+    setDescription("Spark Driver for ETL Batch Applications");
+    setMainClass(ETLSparkProgram.class);
+
+    PipelineRegisterer pipelineRegisterer = new PipelineRegisterer(getConfigurer(), "batch");
+
+    Pipeline pipelineIds =
+      pipelineRegisterer.registerPlugins(
+        config, TimePartitionedFileSet.class,
+        FileSetProperties.builder()
+          .setInputFormat(AvroKeyInputFormat.class)
+          .setOutputFormat(AvroKeyOutputFormat.class)
+          .setEnableExploreOnCreate(true)
+          .setSerDe("org.apache.hadoop.hive.serde2.avro.AvroSerDe")
+          .setExploreInputFormat("org.apache.hadoop.hive.ql.io.avro.AvroContainerInputFormat")
+          .setExploreOutputFormat("org.apache.hadoop.hive.ql.io.avro.AvroContainerOutputFormat")
+          .setTableProperty("avro.schema.literal", Constants.ERROR_SCHEMA.toString())
+          .build(), true);
+
+    if (config.getResources() != null) {
+      setDriverResources(config.getResources());
+      setExecutorResources(config.getResources());
+    }
+
+    // add source, sink, transform ids to the properties. These are needed at runtime to instantiate the plugins
+    Map<String, String> properties = new HashMap<>();
+    properties.put(Constants.PIPELINEID, GSON.toJson(pipelineIds));
+    setProperties(properties);
+  }
+
+  @Override
+  public void beforeSubmit(SparkContext context) throws Exception {
+    cleanupFiles = new ArrayList<>();
+    finishers = new ArrayList<>();
+
+    Map<String, String> properties = context.getSpecification().getProperties();
+    Pipeline pipeline = GSON.fromJson(properties.get(Constants.PIPELINEID), Pipeline.class);
+    String sourcePluginId = pipeline.getSource();
+
+    PluginContext pluginContext = context.getPluginContext();
+    BatchConfigurable<BatchSourceContext> batchSource = pluginContext.newPluginInstance(sourcePluginId);
+    SparkBatchSourceContext sourceContext = new SparkBatchSourceContext(context,
+                                                                        new DatasetContextLookupProvider(context),
+                                                                        sourcePluginId);
+    batchSource.prepareRun(sourceContext);
+
+    SparkBatchSourceFactory sourceFactory = sourceContext.getSourceFactory();
+    if (sourceFactory == null) {
+      // TODO: Revisit what exception to throw
+      throw new IllegalArgumentException("Input is not set");
```

Comment:
More descriptive errors. `Input for ETLSparkJob is not set. <Specify how to configure input>`